posts/2021-07-21-building-on-atomic-reference.adoc

```diff
@@ -0,0 +1,208 @@
+blog_description: "Java's AtomicReference does not guarantee that an update function gets called exactly once. We describe a helpful, thread-safe mechanism that builds on AtomicReference to provide such a guarantee. This can be used for lazy initialization, clean-up, or any state change processing that is to happen exactly once."
+open-graph-image: https://openliberty.io/img/twitter_card.jpg
+additional_authors:
+- name: Neil Richards
+  github: https://github.com/ngmr
+  image: https://avatars0.githubusercontent.com/ngmr
+---
+= Building on AtomicReference
+Joe Chacko <https://github.com/joe-chacko>
+:imagesdir: /
+:url-prefix:
+:url-about: /
+//Blank line here is necessary before starting the body of the post.
+
+The https://devdocs.io/openjdk/java.base/java/util/concurrent/atomic/atomicreference[`AtomicReference`] and other `Atomic...` classes in the `java.util.concurrent` package allow multiple threads to read and update values safely.
+Yet `AtomicReference` alone is of limited use in writing thread-safe code.
+This article describes a helpful mechanism to provide a more useful thread-safe reference.
+
+== Something is rotten in the state transition
+Consider the https://devdocs.io/openjdk~15/java.base/java/util/concurrent/atomic/atomicreference#updateAndGet(java.util.function.UnaryOperator)[`updateAndGet()`] method.
+This allows many threads to contend to update the value.
+For example, one might write something like this:
+[source, java]
+----
+    private AtomicReference<Widget> ref;
+
+    public Widget getWidget() {
+        return ref.updateAndGet(value -> value == null ? new Widget() : value);
+    }
+----
+When `getWidget()` is first invoked, it will create a new `Widget`, but later calls will find the existing one.
+
+But there is a trap here for the unwary developer.
+The contract of the `updateAndGet()` method describes some restrictions on the update function: --
+[quote]
+----
+"The function should be side-effect-free, since it may be re-applied when attempted updates fail due to contention among threads."
+----
+Although the update only happens once, many threads may call the update function at the same time.
+If the update function is fast and side-effect free, there is no cause for concern.
+But if creating a new widget is slow, or has side-effects, this approach is flawed.
+Let's build an alternative that works well with any kind of update function,
+even if it is slow, non-thread-safe, or throws exceptions.
+
+== To do or not to do
+The update processing  --  e.g. a lazy initialization step  --  should happen exactly once.
+Thread contention must be resolved _before the work is undertaken_: --
+
+* Exactly one thread must win the race *to do* the update.
+* All other threads need *not to do* anything but wait for the result.
+
+Let's build a new utility.
+We'll call it `AcidReference`.
+
+.What's in a name?
```

Comment:
I see this is styled as a callout, but no special styling is applied on the actual site. I know you and Steven were working on CSS adjustments, but if they won't be applied by the time this publishes, I recommend replacing the callout with an h3 section, or rewriting to incorporate with the previous section. Similar recommendation for the other callouts in the post. The content itself is fine but if the styling is not applied, the structure looks weird and arbitrary.

Reply:
Disregard this comment- I see the callouts on the draft site now

Reply:
Disregarded.